topi/python/topi/x86/conv2d.py

```diff
@@ -462,8 +699,212 @@ def traverse(op):
+
+    # Inline the constants.
+    s[A].compute_inline()
+    s[B].compute_inline()
+
+    # transform kernel
+    if isinstance(U.op, tvm.tensor.ComputeOp):
+        kernel, G = U.op.input_tensors
+        s[G].compute_inline()
+        coo, cii, eps, nu, ciii, vc = s[U].op.axis
+        if autotvm.GLOBAL_SCOPE.in_tuning:
+            # kernel transformation will be pre-computed during compilation, so we skip
+            # this part to make tuning records correct
+            s[U].pragma(eps, 'debug_skip_region')
+        else:
+            pass
+            # r_kh, r_kw = s[U].op.reduce_axis
+            # s[U].reorder(k, c, eps, nu, r_kh, r_kw, kk)
+            # for axis in [eps, nu, r_kh, r_kw]:
+            #     s[U].unroll(axis)
+            # s[U].vectorize(kk)
+            # s[U].parallel(k)
+
+        if isinstance(kernel.op, tvm.tensor.ComputeOp) and "dilate" in kernel.op.tag:
+            s[kernel].compute_inline()
+
+    ############################################################
+    # input tile
+    n, cii, oh_m, ow_m, eps, nu, ciii = s[input_tile].op.axis
+    # Vectorize the input tile
+    s[input_tile].vectorize(ciii)
+
+    cfg.define_knob('data_pad_compute_location', [0, 1, 2, 3])
+    if cfg['data_pad_compute_location'].val == 0:
+        parallel_axis = s[input_tile].fuse(n)
+        s[data_pad].compute_inline()
+    if cfg['data_pad_compute_location'].val == 1:
+        parallel_axis = s[input_tile].fuse(n)
+        s[data_pad].compute_at(s[input_tile], cii)
+        (_, _, _, _, dpcii) = s[data_pad].op.axis
+        s[data_pad].vectorize(dpcii)
+    if cfg['data_pad_compute_location'].val == 2:
+        parallel_axis = s[input_tile].fuse(n, cii)
+        s[data_pad].compute_at(s[input_tile], oh_m)
+        (_, _, _, _, dpcii) = s[data_pad].op.axis
+        s[data_pad].vectorize(dpcii)
+    if cfg['data_pad_compute_location'].val == 3:
+        parallel_axis = s[input_tile].fuse(n, cii, oh_m)
+        s[data_pad].compute_at(s[input_tile], ow_m)
+        (_, _, _, _, dpcii) = s[data_pad].op.axis
+        s[data_pad].vectorize(dpcii)
+
+    # s[input_tile].parallel(parallel_axis)
+    ############################################################
+
+    ############################################################
+    # data_pad
+    # s[data_pad].compute_inline()
+    ############################################################
+
+    ############################################################
+    # transform image
+    n, cii, oh_m, ow_m, eps, nu, ciii = s[V].op.axis
+    r_eps, r_nu = s[V].op.reduce_axis
+
+    s[V].vectorize(ciii)
+    cfg["reorder_V"].apply(s, V, [n, cii, oh_m, ow_m, eps, nu, ciii, r_eps, r_nu])
+
+    cfg.define_annotate("reduce_V", [r_eps, r_nu, eps, nu],
+                        policy='unroll')
+    cfg['reduce_V'].apply(s, V, [r_eps, r_nu, eps, nu], cfg=cfg)
+
+
+    cfg.define_knob('input_tile_compute_location', [0, 1, 2, 3])
+    if cfg['input_tile_compute_location'].val == 0:
```

Comment:
@merrymercy ,

I spent some time to see how to replace ```knob``` with something more elegant. They are used in more places, operations involved by ```knob``` are different, don't see how can generalize.

* ```knob``` directs ```compute_at``` & ```vectorize``` but also ```fuse``` & ```compute_at``` too.
* If have ideas please elaborate, otherwise can we leave as ```knob``` ?